content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,357 @@
+[Git, popular OSS project](https://public-inbox.org/git/CAGm8dMApDdLEzeKU-h16G0NSpnuk9LMTWA29t4MxO1qcNpUvhA@mail.gmail.com/).
+
+### 3. 2011 Code review integration
+
+Here the intuition was that at the code review stage developers are preparing their code to be seen, so they
+are more receptive to suggestions and readability and stylistic comments.
+
+This is aligned with the direction that few people at [source{d}](https://medium.com/@sourcedtech) are
+[exploring now](https://github.com/campoy/goodgopher/blob/d15a7eccc3bcb2484234732e538fc425d728e972/README.md).
+There is also some number of existing solutions in the same field:
+
+* [https://codeclimate.com](https://codeclimate.com/)
+* [https://www.codacy.com](https://www.codacy.com/)
+* [https://sideci.com](https://sideci.com/)
+* [https://houndci.com](https://houndci.com/)
+* [https://github.com/haya14busa/reviewdog](https://github.com/haya14busa/reviewdog)
+
+Authors were focused on just presenting the FindBugs results at the internal code review tool as “comments”,
+but this attempt did not take off either. Their integration included a few notable features:
+
+* an individual developer could suppress false positives and filter the comments by confidence levels
+* it was meant to show only new FindBugs warnings, but sometimes those issues were miscategorized as new
+
+Although code review still seems to be the best time for surfacing analysis results, this attempt failure was
+attributed to:
+
+* presence of false positives in FindBugs results made developers lose confidence in the tool as a whole
+* customization of the results view per-developer led to an inconsistent view of the analysis outcome
+
+## What worked & lessons learned
+
+As opposed to integrating each particular analysis tool in a different way, an internal “platform” — easily
+extensible and with support for many different kinds of program-analysis tools, including static and dynamic
+analyses was built, known as a [Tricorder project](https://research.google.com/pubs/pub43322.html).
+
+As it was taking into account all the lessons learned from the history above, it managed to regain the trust of
+the users and proved to be a success inside Google.
+
+The paper contains a number of general lessons like *Developer happiness is key* and *Crowdsource analysis development*
+that are nice, but I would rather highlight a few key takeaways instead, that seem to drive the rest of
+the technical decisions, responsible for the success of a new analysis platform.
+
+There are two main takeaways that drove the overall tooling design:
+
+### 1. Best way to **measure a success of analysis**
+> by number of bugs fixed (or prevented), not the number of issues identified
+
+This way of measuring a success has several notable implications:
+
+* If the tool that finds a bug also suggests a fix - it will be much more successful by this metric.
+  This by necessity constraints the scope of a possible analysis and a tooling required
```

Comment:
any possible analysis and the tooling required.

(remember to finish all sentences with `.`)